controllers/registry_controller.go

```diff
@@ -71,9 +85,118 @@ func (r *registryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (r
 		err = markAsFreeOnSuccess(r.client, r.watchNamespace, r.recorder, RegistryController, result.Requeue, err)
 	}()
 
+	// List all IDMS/ITMS resources
+	imageDigestMirrorSetList := &config.ImageDigestMirrorSetList{}
+	if err = r.client.List(ctx, imageDigestMirrorSetList); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting IDMS list: %w", err)
+	}
+	imageTagMirrorSetList := &config.ImageTagMirrorSetList{}
+	if err = r.client.List(ctx, imageTagMirrorSetList); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting ITMS list: %w", err)
+	}
+
+	mergedMirrorSets := collectMirrorSets(imageDigestMirrorSetList.Items, imageTagMirrorSetList.Items)
+
 	return ctrl.Result{}, nil
 }
 
+// collectMirrorSets extracts the contents of the given mirror sets and merges them into a single collection.
+// Each member of the collection represents the registry configuration for a specific source.
+func collectMirrorSets(idmsItems []config.ImageDigestMirrorSet, idtsItems []config.ImageTagMirrorSet) []mirrorSet {
+	var allMirrorSets []mirrorSet
+	for _, idms := range idmsItems {
+		for _, entry := range idms.Spec.ImageDigestMirrors {
+			set := &mirrorSet{
+				source:             entry.Source,
+				mirrorSourcePolicy: entry.MirrorSourcePolicy,
+			}
+			for _, image := range entry.Mirrors {
+				set.mirrors = append(set.mirrors, mirror{host: string(image), resolveTags: false})
+			}
+			allMirrorSets = append(allMirrorSets, *set)
+		}
+	}
+	for _, itms := range idtsItems {
+		for _, entry := range itms.Spec.ImageTagMirrors {
+			set := &mirrorSet{
+				source:             entry.Source,
+				mirrorSourcePolicy: entry.MirrorSourcePolicy,
+			}
+			for _, image := range entry.Mirrors {
+				set.mirrors = append(set.mirrors, mirror{host: string(image), resolveTags: true})
+			}
+			allMirrorSets = append(allMirrorSets, *set)
+		}
+	}
+	return mergeMirrorSets(allMirrorSets)
+}
+
+// mergeMirrorSets consolidates duplicate entries in the given slice (based on the source) since we do not want to
+// generate multiple config files for the same source image repo. Output is sorted to ensure it is deterministic.
+func mergeMirrorSets(baseMirrorSets []mirrorSet) []mirrorSet {
+	// Map to keep track of unique mirrorSets by source
+	uniqueMirrorSets := make(map[string]mirrorSet)
+
+	for _, ms := range baseMirrorSets {
+		if existingMS, ok := uniqueMirrorSets[ms.source]; ok {
+			// If the source already exists, merge its mirrors slices and update the map entry
+			existingMS.mirrors = mergeMirrors(existingMS.mirrors, ms.mirrors)
```

Comment:
what if the mirror sets have conflicting a mirrorSourcePolicy?

Reply:
Good point, I looked at what MCO does and it looks like "NeverContactSource" wins out if there is conflict. Erring on the side of caution security wise

Reply:
Don't think so. Here's the package they use for merging and editing registry config https://pkg.go.dev/github.com/openshift/runtime-utils/pkg/registries#EditRegistriesConfig